parser_library/test/lsp/lsp_context_preprocessor_test.cpp

```diff
@@ -0,0 +1,113 @@
+} // namespace
+
+class lsp_context_endevor_preprocessor_test : public testing::Test
+{
+public:
+    lsp_context_endevor_preprocessor_test()
+        : lib_provider({ { "MEMBER", R"(R2 EQU 2
+            LR R2,R2)" },
+            { "MEMBER2", R"(R5 EQU 5
+            LR R5,R5)" } })
+        , a(contents, analyzer_options { source_loc, &lib_provider, endevor_preprocessor_options() }) {};
+
+    void SetUp() override { a.analyze(); }
+    void TearDown() override {}
+
+protected:
+    const std::string contents =
+        R"(
+-INC  MEMBER blabla
+++INCLUDE  MEMBER blabla
+-INC  MEMBER2 blabla
+)";
+    std::string content;
+    mock_parse_lib_provider lib_provider;
+    analyzer a;
+};
+
+TEST_F(lsp_context_endevor_preprocessor_test, go_to)
+{
+    // no jump, instr -INC
+    EXPECT_EQ(location(position(1, 1), source_loc), a.context().lsp_ctx->definition(source_loc, position(1, 1)));
+    // no jump, instr ++INCLUDE
+    EXPECT_EQ(location(position(2, 5), source_loc), a.context().lsp_ctx->definition(source_loc, position(2, 5)));
+    // no jump, instr -INC
+    EXPECT_EQ(location(position(3, 1), source_loc), a.context().lsp_ctx->definition(source_loc, position(3, 1)));
+
+    // jump from source to included file
+    EXPECT_EQ(location(position(0, 3), member_loc), a.context().lsp_ctx->definition(source_loc, position(1, 8)));
+    // jump from source to included file
+    EXPECT_EQ(location(position(0, 3), member_loc), a.context().lsp_ctx->definition(source_loc, position(2, 14)));
+    // jump from source to included file
+    EXPECT_EQ(location(position(0, 3), member2_loc), a.context().lsp_ctx->definition(source_loc, position(3, 8)));
+
+    // no jump
+    EXPECT_EQ(location(position(1, 15), source_loc), a.context().lsp_ctx->definition(source_loc, position(1, 15)));
+    // no jump
+    EXPECT_EQ(location(position(2, 21), source_loc), a.context().lsp_ctx->definition(source_loc, position(2, 21)));
+    // no jump
+    EXPECT_EQ(location(position(3, 15), source_loc), a.context().lsp_ctx->definition(source_loc, position(3, 15)));
+}
+
+TEST_F(lsp_context_endevor_preprocessor_test, refs)
+{
+    location_list expected_inc_locations { location(position(1, 0), source_loc), location(position(3, 0), source_loc) };
+    location_list expected_include_locations { location(position(2, 0), source_loc) };
+    location_list expected_member_locations { location(position(1, 6), source_loc),
+        location(position(2, 11), source_loc) };
+    location_list expected_member2_locations { location(position(3, 6), source_loc) };
+    location_list expected_blabla_locations {};
+
+    // -INC reference
+    EXPECT_EQ(expected_inc_locations, a.context().lsp_ctx->references(source_loc, position(1, 1)));
+    // ++INCLUDE reference
+    EXPECT_EQ(expected_include_locations, a.context().lsp_ctx->references(source_loc, position(2, 5)));
+    // -INC reference
+    EXPECT_EQ(expected_inc_locations, a.context().lsp_ctx->references(source_loc, position(3, 2)));
```

Comment:
Just an idea to consider - should we consider -INC and ++INCLUDE to be the same entity and therefore return references to both of them?

Reply:
I thought about this too and I was undecided on this... I guess some user who knows only about -INC and doesn't know about ++INCLUDE might be surprised to see it referenced at first but it might ultimately be handy to him. 

I'll change it and let's see if there are any reactions from actual users